File: python/1.fibbonaci.py
#!/usr/bin/env python3
import sys
import sysconfig
import time
from threading import Thread
from multiprocessing import Process


def compute_fibonacci(n):
    a, b = 0, 1
    for _ in range(n):
        a, b = b, a + b
    return a


def run_single_threaded(nums):
    for num in nums:
        compute_fibonacci(num)


def run_multi_threaded(nums):
    threads = [Thread(target=compute_fibonacci, args=(num,)) for num in nums]
    for thread in threads:
        thread.start()
    for thread in threads:
        thread.join()


def run_multi_processing(nums):
    processes = [Process(target=compute_fibonacci, args=(num,)) for num in nums]
    for process in processes:
        process.start()
    for process in processes:
        process.join()


def main():
    print(f"Python Version: {sys.version}")

    py_version = float(".".join(sys.version.split()[0].split(".")[:2]))
    status = sysconfig.get_config_var("Py_GIL_DISABLED")

    if py_version >= 3.13:
        status = sys._is_gil_enabled()

    if status is None:
        print("GIL cannot be disabled for Python <= 3.12")
    elif status == 0:
        print("GIL is currently disabled")
    elif status == 1:
        print("GIL is currently active")

    nums = [300000] * 10

    print("\nRunning Single-Threaded Task:")
    start = time.perf_counter()
    run_single_threaded(nums)
    print(f"run_single_threaded took {time.perf_counter() - start:.4f} seconds.")

    print("\nRunning Multi-Threaded Task:")
    start = time.perf_counter()
    run_multi_threaded(nums)
    print(f"run_multi_threaded took {time.perf_counter() - start:.4f} seconds.")

    print("\nRunning Multi-Processing Task:")
    start = time.perf_counter()
    run_multi_processing(nums)
    print(f"run_multi_processing took {time.perf_counter() - start:.4f} seconds.")


if __name__ == "__main__":
    main()

File: go/1.fibbonaci.go
package main

import (
	"fmt"
	"math/big"
	"runtime"
	"sync"
	"time"
)

func measureExecutionTime(name string, fn func()) {
	start := time.Now()
	fn()
	elapsed := time.Since(start)
	fmt.Printf("%s took %.4f seconds.\n", name, elapsed.Seconds())
}

func computeFibonacci(n int) *big.Int {
	a := big.NewInt(0)
	b := big.NewInt(1)
	temp := new(big.Int)


	for i := 0; i < n; i++ {
		temp.Set(a)
		a.Set(b)
		b.Add(temp, b)
	}
	return a
}

func runSingleThreaded(nums []int) {
	for _, num := range nums {
		computeFibonacci(num)
	}
}

func runMultiThreaded(nums []int) {
	var wg sync.WaitGroup
	wg.Add(len(nums))

	for _, num := range nums {
		go func(n int) {
			defer wg.Done()
			computeFibonacci(n)
		}(num)
	}

	wg.Wait()
}

func main() {
	fmt.Printf("Go Version: %s\n", runtime.Version())
	fmt.Printf("GOMAXPROCS: %d (number of CPUs available)\n", runtime.GOMAXPROCS(0))
	fmt.Printf("NumCPU: %d\n", runtime.NumCPU())

	fmt.Println("\nNote: Go has no GIL - goroutines execute in true parallelism")

	nums := make([]int, 10)
	for i := range nums {
		nums[i] = 300000
	}

	fmt.Println("\nRunning Single-Threaded Task:")
	measureExecutionTime("runSingleThreaded", func() {
		runSingleThreaded(nums)
	})

	fmt.Println("\nRunning Multi-Threaded Task (Goroutines):")
	measureExecutionTime("runMultiThreaded", func() {
		runMultiThreaded(nums)
	})

	fmt.Println("\nNote: Go goroutines already provide true parallelism (no separate multiprocessing needed)")
}
